File: src/utils/scoringEngine.ts
import { AnswerRecord, Question, CompetencyScore } from '../types';

export interface EvaluationResult {
  totalScore: number;
  recommendation: 'Strong Hire' | 'Hire' | 'Borderline' | 'No Hire';
  recommendationRationale: string;
  competencies: {
    technicalKnowledge: number;
    problemSolving: number;
    communication: number;
    architectureSkills: number;
    codingSkills: number;
  };
  strengths: string[];
  weaknesses: string[];
  keyObservations: string[];
  areasForImprovement: string[];
  overallSummary: string;
}

export function evaluateInterview(
  role: string,
  candidateName: string,
  answers: AnswerRecord[],
  questions: Question[],
  interviewerNotes?: string
): EvaluationResult {
  if (!answers || answers.length === 0) {
    return {
      totalScore: 0,
      recommendation: 'No Hire',
      recommendationRationale: 'No questions were evaluated.',
      competencies: {
        technicalKnowledge: 1,
        problemSolving: 1,
        communication: 1,
        architectureSkills: 1,
        codingSkills: 1
      },
      strengths: ['Interview in progress'],
      weaknesses: ['Evaluation incomplete'],
      keyObservations: ['Session started'],
      areasForImprovement: ['Complete all interview questions'],
      overallSummary: 'The interview is currently in progress.'
    };
  }

  // Calculate raw scores and competency breakdowns
  let earnedScore = 0;
  let maxPossibleScore = 0;

  const competencyTotals: Record<string, { earned: number; max: number; count: number }> = {
    'Technical Knowledge': { earned: 0, max: 0, count: 0 },
    'Problem Solving': { earned: 0, max: 0, count: 0 },
    'Communication': { earned: 0, max: 0, count: 0 },
    'Architecture Skills': { earned: 0, max: 0, count: 0 },
    'Coding Skills': { earned: 0, max: 0, count: 0 }
  };

  const strengthsList: string[] = [];
  const weaknessesList: string[] = [];

  answers.forEach(ans => {
    const q = questions.find(item => item.questionId === ans.questionId);
    const qMax = q ? q.score : 10;
    const qComp = q ? q.competency : (ans.competency || 'Technical Knowledge');

    earnedScore += ans.scoreAwarded;
    maxPossibleScore += qMax;

    if (!competencyTotals[qComp]) {
      competencyTotals[qComp] = { earned: 0, max: 0, count: 0 };
    }
    competencyTotals[qComp].earned += ans.scoreAwarded;
    competencyTotals[qComp].max += qMax;
    competencyTotals[qComp].count += 1;

    if (ans.copilotFeedback) {
      if (ans.copilotFeedback.strength && ans.scoreAwarded >= 7) {
        strengthsList.push(ans.copilotFeedback.strength);
      }
      if (ans.copilotFeedback.weakness && ans.scoreAwarded < 9 && !ans.copilotFeedback.weakness.includes('None')) {
        weaknessesList.push(ans.copilotFeedback.weakness);
      }
    } else if (ans.isCorrect) {
      strengthsList.push(`Accurately answered ${q?.difficulty || 'core'} ${qComp} question.`);
    } else if (ans.isCorrect === false) {
      weaknessesList.push(`Missed key concept in ${qComp}: ${q?.question.substring(0, 60)}...`);
    }
  });

  const normalizedPercentage = maxPossibleScore > 0 
    ? Math.round((earnedScore / maxPossibleScore) * 100) 
    : 0;

  // Compute 1-5 scale for each competency
  const mapToFiveScale = (compName: string, fallbackDefault: number): number => {
    const data = competencyTotals[compName];
    if (!data || data.max === 0) return fallbackDefault;
    const ratio = data.earned / data.max;
    // Scale 0..1 to 1..5
    return Math.min(5, Math.max(1, Math.round(ratio * 4 + 1)));
  };

  const compScores = {
    technicalKnowledge: mapToFiveScale('Technical Knowledge', Math.max(1, Math.round(normalizedPercentage / 20))),
    problemSolving: mapToFiveScale('Problem Solving', Math.max(1, Math.round((normalizedPercentage * 0.95) / 20))),
    communication: mapToFiveScale('Communication', 4), // Default solid communication in simulated interview
    architectureSkills: mapToFiveScale('Architecture Skills', Math.max(1, Math.round((normalizedPercentage * 0.9) / 20))),
    codingSkills: mapToFiveScale('Coding Skills', Math.max(1, Math.round(normalizedPercentage / 20)))
  };

  // Rule-based Recommendation
  let recommendation: 'Strong Hire' | 'Hire' | 'Borderline' | 'No Hire';
  let recommendationRationale: string;

  if (normalizedPercentage >= 85) {
    recommendation = 'Strong Hire';
    recommendationRationale = `Candidate achieved a stellar score of ${normalizedPercentage}%, demonstrating deep technical mastery and clear architectural articulation for the ${role} position. Highly recommended for immediate offer.`;
  } else if (normalizedPercentage >= 70) {
    recommendation = 'Hire';
    recommendationRationale = `Candidate demonstrated solid core competency with an overall score of ${normalizedPercentage}%. Possesses strong execution abilities with minor coaching opportunities in high-scale scenarios.`;
  } else if (normalizedPercentage >= 50) {
    recommendation = 'Borderline';
    recommendationRationale = `Candidate scored ${normalizedPercentage}%, showing basic foundational knowledge but notable gaps in advanced problem solving and architecture. Recommend a follow-up panel or leveling discussion.`;
  } else {
    recommendation = 'No Hire';
    recommendationRationale = `Candidate scored ${normalizedPercentage}%, falling below the benchmark threshold for the ${role} role. Critical competencies were not demonstrated.`;
  }

  // Deduplicate strengths/weaknesses
  const finalStrengths = Array.from(new Set(strengthsList)).slice(0, 4);
  if (finalStrengths.length === 0) {
    finalStrengths.push(`Solid grasp of fundamental ${role} principles.`);
  }

  const finalWeaknesses = Array.from(new Set(weaknessesList)).slice(0, 3);
  if (finalWeaknesses.length === 0) {
    finalWeaknesses.push('No significant technical deficiencies identified.');
  }

  const keyObservations = [
    `${candidateName} demonstrated ${normalizedPercentage >= 70 ? 'strong confidence' : 'hesitation'} when defending technical trade-offs.`,
    `Performance was consistently high in ${Object.entries(compScores).sort((a, b) => b[1] - a[1])[0][0].replace(/([A-Z])/g, ' $1').toLowerCase()}.`,
    interviewerNotes ? `Interviewer Note: ${interviewerNotes}` : 'Engaged constructively with copilot follow-up prompts.'
  ];

  const areasForImprovement = finalWeaknesses.map(w => `Focus on: ${w}`);

  const overallSummary = `${candidateName} completed the structured ${role} evaluation with an aggregate score of ${normalizedPercentage}/100 (${recommendation}). ${
    recommendation === 'Strong Hire' 
      ? 'The candidate excels in complex problem solving and modern engineering patterns with exceptional clarity.'
      : recommendation === 'Hire'
      ? 'The candidate demonstrates dependable practical abilities and meets all baseline technical requirements.'
      : recommendation === 'Borderline'
      ? 'The candidate demonstrates basic literacy but struggled with senior-level architectural constraints.'
      : 'The candidate failed to demonstrate sufficient foundational competency for this role.'
  }`;

  return {
    totalScore: normalizedPercentage,
    recommendation,
    recommendationRationale,
    competencies: compScores,
    strengths: finalStrengths,
    weaknesses: finalWeaknesses,
    keyObservations,
    areasForImprovement,
    overallSummary
  };
}

File: app/services/scoring_engine.py
from typing import List, Dict, Optional, Tuple
from app.models.interview import AnswerRecord, EvaluationResult, CompetencyScores
from app.models.question import Question

def evaluate_interview_answers(
    role: str,
    candidate_name: str,
    answers: List[AnswerRecord],
    questions: List[Question],
    interviewer_notes: str = ""
) -> EvaluationResult:
    """
    Pure deterministic rule-based evaluation engine.
    Calculates competency breakdown (1-5), aggregate score (0-100),
    and hiring recommendation based on strict business logic.
    """
    if not answers:
        return EvaluationResult(
            totalScore=0,
            recommendation="No Hire",
            recommendationRationale="No questions were evaluated.",
            competencies=CompetencyScores(
                technicalKnowledge=1,
                problemSolving=1,
                communication=1,
                architectureSkills=1,
                codingSkills=1
            ),
            strengths=["Interview in progress"],
            weaknesses=["Evaluation incomplete"],
            keyObservations=["Session started"],
            areasForImprovement=["Complete all interview questions"],
            overallSummary="The interview is currently in progress."
        )

    q_map: Dict[str, Question] = {q.questionId: q for q in questions}

    earned_score = 0
    max_score = 0

    comp_totals = {
        "Technical Knowledge": {"earned": 0, "max": 0},
        "Problem Solving": {"earned": 0, "max": 0},
        "Communication": {"earned": 0, "max": 0},
        "Architecture Skills": {"earned": 0, "max": 0},
        "Coding Skills": {"earned": 0, "max": 0},
    }

    strengths_list = []
    weaknesses_list = []

    for ans in answers:
        q = q_map.get(ans.questionId)
        q_max = q.score if q else 10
        q_comp = q.competency if q else ans.competency

        earned_score += ans.scoreAwarded
        max_score += q_max

        if q_comp not in comp_totals:
            comp_totals[q_comp] = {"earned": 0, "max": 0}

        comp_totals[q_comp]["earned"] += ans.scoreAwarded
        comp_totals[q_comp]["max"] += q_max

        if ans.copilotFeedback:
            if ans.copilotFeedback.strength and ans.scoreAwarded >= 7:
                strengths_list.append(ans.copilotFeedback.strength)
            if ans.copilotFeedback.weakness and ans.scoreAwarded < 9 and "None" not in ans.copilotFeedback.weakness:
                weaknesses_list.append(ans.copilotFeedback.weakness)
        elif ans.isCorrect:
            strengths_list.append(f"Accurately answered core {q_comp} question.")
        elif ans.isCorrect is False:
            q_text = q.question[:60] if q else "concept"
            weaknesses_list.append(f"Missed key concept in {q_comp}: {q_text}...")

    normalized_percentage = round((earned_score / max_score) * 100) if max_score > 0 else 0

    def calc_1_to_5_score(comp_name: str, fallback: int) -> int:
        data = comp_totals.get(comp_name)
        if not data or data["max"] == 0:
            return fallback
        ratio = data["earned"] / data["max"]
        return min(5, max(1, round(ratio * 4 + 1)))

    tech_score = calc_1_to_5_score("Technical Knowledge", max(1, round(normalized_percentage / 20)))
    prob_score = calc_1_to_5_score("Problem Solving", max(1, round((normalized_percentage * 0.95) / 20)))
    comm_score = calc_1_to_5_score("Communication", 4)
    arch_score = calc_1_to_5_score("Architecture Skills", max(1, round((normalized_percentage * 0.9) / 20)))
    code_score = calc_1_to_5_score("Coding Skills", max(1, round(normalized_percentage / 20)))

    # Recommendation Rules
    if normalized_percentage >= 85:
        recommendation = "Strong Hire"
        recommendation_rationale = (
            f"Candidate achieved a standout score of {normalized_percentage}%, demonstrating deep technical "
            f"mastery and architectural reasoning for the {role} role. Highly recommended for immediate offer."
        )
    elif normalized_percentage >= 70:
        recommendation = "Hire"
        recommendation_rationale = (
            f"Candidate demonstrated solid technical fundamentals with an overall score of {normalized_percentage}%. "
            f"Possesses strong execution abilities with minor coaching opportunities in high-scale scenarios."
        )
    elif normalized_percentage >= 50:
        recommendation = "Borderline"
        recommendation_rationale = (
            f"Candidate scored {normalized_percentage}%, showing acceptable foundational literacy but noticeable "
            f"gaps in advanced problem solving and architecture. Recommend a leveling discussion."
        )
    else:
        recommendation = "No Hire"
        recommendation_rationale = (
            f"Candidate scored {normalized_percentage}%, falling below the benchmark threshold for the {role} role. "
            f"Critical competencies were not demonstrated."
        )

    # Deduplicate strengths / weaknesses
    final_strengths = list(dict.fromkeys(strengths_list))[:4]
    if not final_strengths:
        final_strengths.append(f"Solid grasp of fundamental {role} engineering concepts.")

    final_weaknesses = list(dict.fromkeys(weaknesses_list))[:3]
    if not final_weaknesses:
        final_weaknesses.append("No significant technical deficiencies identified.")

    key_observations = [
        f"{candidate_name} demonstrated {'high confidence' if normalized_percentage >= 70 else 'hesitation'} when defending technical trade-offs.",
        f"Strongest demonstrated competency was in {role} execution methodologies.",
        f"Interviewer Note: {interviewer_notes}" if interviewer_notes else "Engaged constructively with copilot follow-up prompts."
    ]

    areas_for_improvement = [f"Focus on: {w}" for w in final_weaknesses]

    narrative = (
        f"{candidate_name} completed the structured {role} assessment with an aggregate score of "
        f"{normalized_percentage}/100 ({recommendation}). "
    )
    if recommendation == "Strong Hire":
        narrative += "The candidate excels in complex problem solving and modern engineering patterns with exceptional clarity."
    elif recommendation == "Hire":
        narrative += "The candidate demonstrates dependable practical abilities and meets all baseline technical requirements."
    elif recommendation == "Borderline":
        narrative += "The candidate demonstrates basic literacy but struggled with senior-level architectural constraints."
    else:
        narrative += "The candidate failed to demonstrate sufficient foundational competency for this role."

    return EvaluationResult(
        totalScore=normalized_percentage,
        recommendation=recommendation,
        recommendationRationale=recommendation_rationale,
        competencies=CompetencyScores(
            technicalKnowledge=tech_score,
            problemSolving=prob_score,
            communication=comm_score,
            architectureSkills=arch_score,
            codingSkills=code_score
        ),
        strengths=final_strengths,
        weaknesses=final_weaknesses,
        keyObservations=key_observations,
        areasForImprovement=areas_for_improvement,
        overallSummary=narrative
    )
